Improve comparability

lcobucci opened the pull request:
I'm working on a new strategy for FastRoute and was using the benchmark to see how good things were. After spending some time on it, I've discovered some minor things that should be improved to allow us to better compare results.

This also fixes #1

lcobucci committed:
Make sure Symfony also validates methods

Aligning to how the other routers work.

lcobucci committed:
Initialise a dummy FastRoute dispatcher for cached stuff

This is also done for the other routers and greatly affects the
comparability of the results.

lcobucci committed:
Add more scenarios to the benchmark

Providing a more complete picture of the solutions.

lcobucci committed:
Use a stable branch of FastRoute

lcobucci committed:
Update README

azjezz commented on README.md:
```diff
@@ -124,87 +118,85 @@ https://github.com/azjezz/benchmark-php-routing/actions
-| symfony:cached(file)                 | benchLast    | 300    | 0.197746 seconds | 1517.0973960914 |
-| symfony:cached(file)                 | benchLongest | 300    | 0.215139 seconds | 1394.447879888  |
-+--------------------------------------+--------------+--------+------------------+-----------------+
++------------------------------ Benchmark results for group cached. -------------+-----------------+
```
can you update with the result from github actions? this looks quite different.

lcobucci replied:
I surely can the only issue is that the result from GHA isn't reliable atm. Opcache optimisations don't kick in if the cache files aren't generated on the same process.

I can add a commit to have a dummy run (quick and dirty solution) before the real one.

azjezz replied:
yep, let's do that

azjezz replied:
maybe a new flag like --warmup

lcobucci replied:
Oops didn't see your comment before pushing, sorry.

https://github.com/azjezz/benchmark-php-routing/pull/4/checks?check_run_id=3003594776 now shows opcache as the fastest option (vs https://github.com/azjezz/benchmark-php-routing/pull/4/checks?check_run_id=3003413955).

I'll update the README.

Would you prefer to have the script changed to use that `--warmup` flag?

lcobucci replied:
README updated.
The results are far different from what I get locally, though...

lcobucci committed:
Make sure cache files exist before running benchmark

Opcache on CLI requires the cache files to be created by a different
process to really kick in. Without that we see very to little benefit
in the file cache, which is actually the fastest strategy on real
servers.

lcobucci commented on README.md:
```diff
@@ -124,87 +118,85 @@ https://github.com/azjezz/benchmark-php-routing/actions
++------------------------------ Benchmark results for group cached. -------------+-----------------+
+| Case                          | Scenario           | Routes | Time             | Per Second      |
++-------------------------------+--------------------+--------+------------------+-----------------+
+| fast-route(mark):cached(file) | benchAll           | 364    | 0.004870 seconds | 74744.279643592 |
```
In my laptop SF router handles this in 0.002404 seconds vs 0.003114 seconds of FastRoute

azjezz replied:
i would consider this to be more consistent, but we can investigate that later, i will do another run locally tomorrow and see
